File: src/lexer.ts
import { err, isalpha, isdigit } from "./utils.ts";

export const enum TokenType {
  OP_ADD,
  OP_SUB,
  OP_MUL,
  OP_DIV,
  OP_NEG,
  OP_POS,
  OP_INC,
  OP_DEC,
  OP_EQ,
  COMP_EQ,
  COMP_NE,
  COMP_GT,
  COMP_GE,
  COMP_LT,
  COMP_LE,
  COMP_AND,
  COMP_OR,
  LOG_NOT,
  LOG_AND,
  LOG_OR,
  LOG_XOR,
  LITERAL,
  LPAREN,
  RPAREN,
  LBRACE,
  RBRACE,
  COLON,
  VAR_DECL,
  VAR_DEL,
  VAR,
  IN,
  OUT,
  IF,
  ELSE,
  WHILE,
  LABEL,
  GOTO,
  EXIT,
  FUNC_DECL,
  CALL,
  RETURN,
  IDENT,
  EOF,
}

export type Literal = string | number | boolean | void;

type Token = {
  type: TokenType;
  name?: string;
  value?: Literal;
};

/**Lexer */
export class Lexer {
  private pos: number;
  private tok: Token;

  constructor(private src: string) {
    this.pos = 0;
    this.tok = { type: TokenType.EOF };
  }

  private current = (): string =>
    this.pos < this.src.length ? this.src[this.pos] : "\0";

  private bump = (): void => {
    this.pos++;
  };

  private skipSpaces = (): void => {
    while (
      this.current() == " " || this.current() == "\t" || this.current() == "\n"
    ) this.bump();
  };

  private parseNumber = (): number => {
    let intPart = 0;
    let fracPart = 0;
    let fracDiv = 1;

    while (isdigit(this.current())) {
      intPart = intPart * 10 + parseInt(this.current());
      this.bump();
    }

    if (this.current() === ".") {
      this.bump();
      if (!isdigit(this.current())) {
        err("Lexer", "Invalid number: expected digit after '.'");
      }
      while (isdigit(this.current())) {
        fracDiv *= 10;
        fracPart = fracPart * 10 + parseInt(this.current());
        this.bump();
      }
    }

    const value = intPart + fracPart / fracDiv;
    return value;
  };

  private parseAlpha = (): string => {
    let alpha: string = "";
    while (isalpha(this.current())) {
      alpha += this.current();
      this.bump();
    }
    return alpha;
  };

  private isPrefix = (): boolean => {
    const prev = this.pos > 0 ? this.src[this.pos - 1] : " ";
    return (
      this.tok.type === TokenType.EOF ||
      this.tok.type === TokenType.LPAREN ||
      this.tok.type === TokenType.OP_EQ ||
      this.tok.type === TokenType.COLON ||
      prev === "=" || prev === "(" || prev === ":"
    );
  };

  public nextToken = (): void => {
    this.skipSpaces();
    if (this.current() === "\0") {
      this.tok = { type: TokenType.EOF };
      return;
    } else if (isdigit(this.current())) {
      const value = this.parseNumber();
      this.tok = { type: TokenType.LITERAL, value };
      return;
    } else if (isalpha(this.current())) {
      const ident = this.parseAlpha();
      switch (ident) {
        case "let": {
          this.tok = { type: TokenType.VAR_DECL };
          break;
        }
        case "out": {
          this.skipSpaces();
          this.tok = { type: TokenType.OUT };
          break;
        }
        case "in": {
          this.skipSpaces();
          this.tok = { type: TokenType.IN };
          break;
        }
        case "true": {
          this.tok = { type: TokenType.LITERAL, value: true };
          break;
        }
        case "false": {
          this.tok = { type: TokenType.LITERAL, value: false };
          break;
        }
        case "null": {
          this.tok = { type: TokenType.LITERAL, value: undefined };
          break;
        }
        case "if": {
          this.tok = { type: TokenType.IF };
          break;
        }
        case "else": {
          this.tok = { type: TokenType.ELSE };
          break;
        }
        case "while": {
          this.tok = { type: TokenType.WHILE };
          break;
        }
        case "goto": {
          this.tok = { type: TokenType.GOTO };
          break;
        }
        case "del": {
          this.tok = { type: TokenType.VAR_DEL };
          break;
        }
        case "exit": {
          this.tok = { type: TokenType.EXIT };
          break;
        }
        case "fun": {
          this.tok = { type: TokenType.FUNC_DECL };
          break;
        }
        case "return": {
          this.tok = { type: TokenType.RETURN };
          break;
        }
        default: {
          this.tok = { type: TokenType.IDENT, name: ident };
          return;
        }
      }
    } else if (this.current() === '"') {
      this.bump();
      let s: string = "";
      while (this.current() !== '"') {
        if (this.current() === "\0") {
          return err("Lexer", "Expected: '\"'");
        }
        s += this.current();
        this.bump();
      }
      this.bump();
      this.tok = { type: TokenType.LITERAL, value: s };
    } else if (this.current() === "'") {
      this.bump();
      let s: string = "";
      while (this.current() !== "'") {
        if (this.current() === "\0") {
          return err("Lexer", 'Expected: "\'"');
        }
        s += this.current();
        this.bump();
      }
      this.bump();
      this.tok = { type: TokenType.LITERAL, value: s };
    } else if (this.current() === "+") {
      if (this.isPrefix()) {
        this.tok = { type: TokenType.OP_POS };
        this.bump();
        return;
      }
      this.bump();
      if (this.current() === "+") {
        this.tok = { type: TokenType.OP_INC };
        this.bump();
        return;
      } else {
        this.tok = { type: TokenType.OP_ADD };
        return;
      }
    } else if (this.current() === "-") {
      if (this.isPrefix()) {
        this.tok = { type: TokenType.OP_NEG };
        this.bump();
        return;
      }
      this.bump();
      if (this.current() === "-") {
        this.tok = { type: TokenType.OP_DEC };
        this.bump();
        return;
      } else {
        this.tok = { type: TokenType.OP_SUB };
        return;
      }
    } else if (this.current() === "*") {
      this.tok = { type: TokenType.OP_MUL };
      this.bump();
      return;
    } else if (this.current() === "/") {
      this.tok = { type: TokenType.OP_DIV };
      this.bump();
      return;
    } else if (this.current() === "(") {
      this.tok = { type: TokenType.LPAREN };
      this.bump();
      return;
    } else if (this.current() === ")") {
      this.tok = { type: TokenType.RPAREN };
      this.bump();
      return;
    } else if (this.current() === "{") {
      this.tok = { type: TokenType.LBRACE };
      this.bump();
      return;
    } else if (this.current() === "}") {
      this.tok = { type: TokenType.RBRACE };
      this.bump();
      return;
    } else if (this.current() === "=") {
      this.bump();
      if (this.current() === "=") {
        this.tok = { type: TokenType.COMP_EQ };
        this.bump();
      } else {
        this.tok = { type: TokenType.OP_EQ };
      }
      return;
    } else if (this.current() === "!") {
      this.bump();
      if (this.current() === "=") {
        this.tok = { type: TokenType.COMP_NE };
        this.bump();
      } else {
        this.tok = { type: TokenType.LOG_NOT };
      }
      return;
    } else if (this.current() === ">") {
      this.bump();
      if (this.current() === "=") {
        this.tok = { type: TokenType.COMP_GE };
        this.bump();
      } else {
        this.tok = { type: TokenType.COMP_GT };
      }
      return;
    } else if (this.current() === "<") {
      this.bump();
      if (this.current() === "=") {
        this.tok = { type: TokenType.COMP_LE };
        this.bump();
      } else {
        this.tok = { type: TokenType.COMP_LT };
      }
      return;
    } else if (this.current() === "&") {
      this.bump();
      if (this.current() === "&") {
        this.tok = { type: TokenType.COMP_AND };
        this.bump();
      } else {
        this.tok = { type: TokenType.LOG_AND };
      }
      return;
    } else if (this.current() === "|") {
      this.bump();
      if (this.current() === "|") {
        this.tok = { type: TokenType.COMP_OR };
        this.bump();
      } else {
        this.tok = { type: TokenType.LOG_OR };
      }
      return;
    } else if (this.current() === "^") {
      this.tok = { type: TokenType.LOG_XOR };
      this.bump();
      return;
    } else if (this.current() === ":") {
      this.tok = { type: TokenType.COLON };
      this.bump();
      return;
    } else if (this.current() === "#") {
      while (this.current() !== "\n" && this.current() !== "\0") this.bump();
      this.nextToken();
      return;
    } else {
      err("Lexer", `Error: Unknown token: '${this.current()}`);
    }
  };

  public currentToken = (): Token => {
    return this.tok;
  };
}

File: src/lexer.rs
use std::{iter::Peekable, str::Chars};

use ordered_float::OrderedFloat;

use crate::token::{Literal, Token, TokenType, VarType};

#[derive(Debug, Clone)]
pub enum LexerError {
    SyntaxError {
        message: String,
        row: usize,
        col: usize,
    },
    InvalidNumber {
        row: usize,
        col: usize,
    },
    UnexpectedChar {
        expected: Option<String>,
        found: char,
        row: usize,
        col: usize,
    },
}

impl std::error::Error for LexerError {}

impl std::fmt::Display for LexerError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            LexerError::SyntaxError { message, row, col } => {
                write!(f, "Syntax error at {}:{}: {}", row, col, message)
            }
            LexerError::InvalidNumber { row, col } => {
                write!(f, "Invalid number at {}:{}", row, col)
            }
            LexerError::UnexpectedChar {
                expected,
                found,
                row,
                col,
            } => {
                if let Some(exp) = expected {
                    write!(
                        f,
                        "Unexpected char at {}:{}: expected '{}', found '{}'",
                        row, col, exp, found
                    )
                } else {
                    write!(f, "Unexpected char at {}:{}: '{}'", row, col, found)
                }
            }
        }
    }
}

#[derive(Debug, Clone)]
pub struct Lexer<'a> {
    tok: Token,
    src: Peekable<Chars<'a>>,
    is_flt: bool,
}

impl<'a> Lexer<'a> {
    pub fn new(src: &'a str) -> Self {
        Lexer {
            tok: Token {
                token: TokenType::EOF,
                value: None,
                row: 1,
                col: 1,
            },
            src: src.chars().peekable(),
            is_flt: false,
        }
    }

    fn current(&mut self) -> char {
        *self.src.peek().unwrap_or(&'\0')
    }

    fn current_safe(&mut self) -> Result<char, LexerError> {
        Ok(*self.src.peek().ok_or_else(|| LexerError::UnexpectedChar {
            expected: None,
            found: '\0',
            row: self.tok.row,
            col: self.tok.col,
        })?)
    }

    fn bump(&mut self) -> () {
        self.src.next();
        self.tok.col += 1;
    }

    fn skip_spaces(&mut self) -> () {
        while self.current() == ' ' || self.current() == '\t' || self.current() == '\n' {
            if self.current() == '\n' {
                self.tok.row += 1;
                self.tok.col = 0;
            }
            self.bump();
        }
    }

    fn parse_number(&mut self) -> Result<f64, LexerError> {
        let mut int_part = 0;
        let mut frac_part = 0;
        let mut frac_div = 1;
        self.is_flt = false;

        while self.current().is_numeric() {
            int_part = int_part * 10
                + self
                    .current()
                    .to_digit(10)
                    .ok_or_else(|| LexerError::InvalidNumber {
                        row: self.tok.row,
                        col: self.tok.col,
                    })?;
            self.bump();
        }

        if self.current() == '.' {
            self.is_flt = true;
            self.bump();
            if !self.current().is_numeric() {
                return Err(LexerError::InvalidNumber {
                    row: self.tok.row,
                    col: self.tok.col,
                });
            }
            while self.current().is_numeric() {
                frac_div *= 10;
                frac_part = frac_part * 10
                    + self
                        .current()
                        .to_digit(10)
                        .ok_or_else(|| LexerError::InvalidNumber {
                            row: self.tok.row,
                            col: self.tok.col,
                        })?;
                self.bump();
            }
        }

        Ok((int_part * frac_div + frac_part) as f64 / frac_div as f64)
    }

    fn parse_ident(&mut self) -> String {
        let mut ident = String::new();

        if self.current().is_ascii_alphabetic() || self.current() == '_' {
            ident.push(self.current());
            self.bump();
        }

        while self.current().is_alphanumeric() || self.current() == '_' {
            ident.push(self.current());
            self.bump();
        }
        ident
    }

    fn is_prefix(&mut self) -> bool {
        let prev = *self.src.peek().unwrap_or(&' ');
        self.tok.token == TokenType::EOF
            || self.tok.token == TokenType::LPAREN
            || self.tok.token == TokenType::EQ
            || prev == '='
            || prev == '('
    }

    pub fn next_token(&mut self) -> Result<(), LexerError> {
        self.skip_spaces();
        if self.current() == '\0' {
            self.tok = Token {
                token: TokenType::EOF,
                value: None,
                row: self.tok.row,
                col: self.tok.col,
            };
            return Ok(());
        } else if self.current().is_numeric() {
            let val = self.parse_number()?;
            if self.is_flt {
                self.tok = Token {
                    token: TokenType::LITERAL(VarType::Float),
                    value: Some(Literal::Float(OrderedFloat(val))),
                    row: self.tok.row,
                    col: self.tok.col,
                };
            } else {
                self.tok = Token {
                    token: TokenType::LITERAL(VarType::Int),
                    value: Some(Literal::Int(val as i64)),
                    row: self.tok.row,
                    col: self.tok.col,
                };
            }
            return Ok(());
        } else if self.current().is_alphabetic() || self.current() == '_' {
            let ident: String = self.parse_ident();
            match ident.as_str() {
                "true" => {
                    self.tok = Token {
                        token: TokenType::LITERAL(VarType::Bool),
                        value: Some(Literal::Bool(true)),
                        row: self.tok.row,
                        col: self.tok.col,
                    };
                }
                "false" => {
                    self.tok = Token {
                        token: TokenType::LITERAL(VarType::Bool),
                        value: Some(Literal::Bool(false)),
                        row: self.tok.row,
                        col: self.tok.col,
                    };
                }
                "null" => {
                    self.tok = Token {
                        token: TokenType::LITERAL(VarType::Void),
                        value: Some(Literal::Void),
                        row: self.tok.row,
                        col: self.tok.col,
                    };
                }
                "let" => {
                    self.tok = Token {
                        token: TokenType::VARDECL,
                        value: None,
                        row: self.tok.row,
                        col: self.tok.col,
                    };
                }
                "fun" => {
                    self.tok = Token {
                        token: TokenType::FUNCDECL,
                        value: None,
                        row: self.tok.row,
                        col: self.tok.col,
                    }
                }
                "return" => {
                    self.tok = Token {
                        token: TokenType::RETURN,
                        value: None,
                        row: self.tok.row,
                        col: self.tok.col,
                    }
                }
                "if" => {
                    self.tok = Token {
                        token: TokenType::IF,
                        value: None,
                        row: self.tok.row,
                        col: self.tok.col,
                    }
                }
                "else" => {
                    self.tok = Token {
                        token: TokenType::ELSE,
                        value: None,
                        row: self.tok.row,
                        col: self.tok.col,
                    }
                }
                "while" => {
                    self.tok = Token {
                        token: TokenType::WHILE,
                        value: None,
                        row: self.tok.row,
                        col: self.tok.col,
                    }
                }
                "goto" => {
                    self.tok = Token {
                        token: TokenType::GOTO,
                        value: None,
                        row: self.tok.row,
                        col: self.tok.col,
                    }
                }
                "extern" => {
                    self.tok = Token {
                        token: TokenType::EXTERN,
                        value: None,
                        row: self.tok.row,
                        col: self.tok.col,
                    }
                }
                "pub" => {
                    self.tok = Token {
                        token: TokenType::PUB,
                        value: None,
                        row: self.tok.row,
                        col: self.tok.col,
                    }
                }
                "int" => {
                    self.tok = Token {
                        token: TokenType::Type(VarType::Int),
                        value: None,
                        row: self.tok.row,
                        col: self.tok.col,
                    }
                }
                "flt" => {
                    self.tok = Token {
                        token: TokenType::Type(VarType::Float),
                        value: None,
                        row: self.tok.row,
                        col: self.tok.col,
                    }
                }
                "str" => {
                    self.tok = Token {
                        token: TokenType::Type(VarType::Str),
                        value: None,
                        row: self.tok.row,
                        col: self.tok.col,
                    }
                }
                "bool" => {
                    self.tok = Token {
                        token: TokenType::Type(VarType::Bool),
                        value: None,
                        row: self.tok.row,
                        col: self.tok.col,
                    }
                }
                "void" => {
                    self.tok = Token {
                        token: TokenType::Type(VarType::Void),
                        value: None,
                        row: self.tok.row,
                        col: self.tok.col,
                    }
                }
                "arr" => {
                    if self.current() != '<' {
                        return Err(LexerError::UnexpectedChar {
                            expected: Some("<".to_string()),
                            found: self.current(),
                            row: self.tok.row,
                            col: self.tok.col,
                        });
                    }
                    self.bump();
                    let len: Option<usize>;
                    if self.current().is_numeric() {
                        len = Some(self.parse_number()? as usize);
                    } else if self.current() == '_' {
                        len = None;
                        self.bump();
                    } else {
                        return Err(LexerError::UnexpectedChar {
                            expected: None,
                            found: self.current(),
                            row: self.tok.row,
                            col: self.tok.col,
                        });
                    }
                    if self.current() != '>' {
                        return Err(LexerError::UnexpectedChar {
                            expected: Some(">".to_string()),
                            found: self.current(),
                            row: self.tok.row,
                            col: self.tok.col,
                        });
                    }
                    self.bump();
                    self.tok = Token {
                        token: TokenType::Type(VarType::Array(len)),
                        value: None,
                        row: self.tok.row,
                        col: self.tok.col,
                    }
                }
                "sizeof" => {
                    self.tok = Token {
                        token: TokenType::SIZEOF,
                        value: None,
                        row: self.tok.row,
                        col: self.tok.col,
                    }
                }
                "for" => {
                    self.tok = Token {
                        token: TokenType::FOR,
                        value: None,
                        row: self.tok.row,
                        col: self.tok.col,
                    }
                }
                "in" => {
                    self.tok = Token {
                        token: TokenType::IN,
                        value: None,
                        row: self.tok.row,
                        col: self.tok.col,
                    }
                }
                _ => {
                    self.tok = Token {
                        token: TokenType::IDENT,
                        value: Some(Literal::Str(ident)),
                        row: self.tok.row,
                        col: self.tok.col,
                    }
                }
            }
            return Ok(());
        } else if self.current() == '"' {
            self.bump();
            let mut s = String::new();
            loop {
                match self.current() {
                    '"' => {
                        self.bump();
                        break;
                    }
                    '\0' => {}
                    '\\' => {
                        self.bump();
                        match self.current() {
                            'n' => {
                                s.push('\n');
                                self.bump();
                            }
                            't' => {
                                s.push('\t');
                                self.bump();
                            }
                            'r' => {
                                s.push('\r');
                                self.bump();
                            }
                            '\\' => {
                                s.push('\\');
                                self.bump();
                            }
                            '"' => {
                                s.push('"');
                                self.bump();
                            }
                            _ => self.bump(),
                        }
                    }
                    c => {
                        s.push(c);
                        self.bump();
                    }
                }
            }
            self.tok = Token {
                token: TokenType::LITERAL(VarType::Str),
                value: Some(Literal::Str(s)),
                row: self.tok.row,
                col: self.tok.col,
            };
            return Ok(());
        } else if self.current() == '\'' {
            self.bump();
            let mut s = String::new();
            while self.current() != '\'' {
                if self.current() == '\0' {
                    return Err(LexerError::UnexpectedChar {
                        expected: Some("'".to_string()),
                        found: self.current(),
                        row: self.tok.row,
                        col: self.tok.col,
                    });
                }
                s.push(self.current());
                self.bump();
            }
            self.bump();
            self.tok = Token {
                token: TokenType::LITERAL(VarType::Str),
                value: Some(Literal::Str(s)),
                row: self.tok.row,
                col: self.tok.col,
            };
            return Ok(());
        } else if self.current() == '+' {
            if self.is_prefix() {
                self.bump();
                return Ok(());
            }
            self.bump();
            if self.current() == '+' {
                self.tok = Token {
                    token: TokenType::INC,
                    value: None,
                    row: self.tok.row,
                    col: self.tok.col,
                };
                self.bump();
                return Ok(());
            }
            self.tok = Token {
                token: TokenType::ADD,
                value: None,
                row: self.tok.row,
                col: self.tok.col,
            };
            return Ok(());
        } else if self.current() == '-' {
            if self.is_prefix() {
                self.tok = Token {
                    token: TokenType::NEG,
                    value: None,
                    row: self.tok.row,
                    col: self.tok.col,
                };
                self.bump();
                return Ok(());
            }
            self.bump();
            if self.current() == '-' {
                self.tok = Token {
                    token: TokenType::DEC,
                    value: None,
                    row: self.tok.row,
                    col: self.tok.col,
                };
                self.bump();
                return Ok(());
            }
            self.tok = Token {
                token: TokenType::SUB,
                value: None,
                row: self.tok.row,
                col: self.tok.col,
            };
            return Ok(());
        } else if self.current() == '*' {
            self.tok = Token {
                token: TokenType::MUL,
                value: None,
                row: self.tok.row,
                col: self.tok.col,
            };
            self.bump();
            return Ok(());
        } else if self.current() == '/' {
            self.tok = Token {
                token: TokenType::DIV,
                value: None,
                row: self.tok.row,
                col: self.tok.col,
            };
            self.bump();
            return Ok(());
        } else if self.current() == '(' {
            self.tok = Token {
                token: TokenType::LPAREN,
                value: None,
                row: self.tok.row,
                col: self.tok.col,
            };
            self.bump();
            return Ok(());
        } else if self.current() == ')' {
            self.tok = Token {
                token: TokenType::RPAREN,
                value: None,
                row: self.tok.row,
                col: self.tok.col,
            };
            self.bump();
            return Ok(());
        } else if self.current() == '{' {
            self.tok = Token {
                token: TokenType::LBRACE,
                value: None,
                row: self.tok.row,
                col: self.tok.col,
            };
            self.bump();
            return Ok(());
        } else if self.current() == '}' {
            self.tok = Token {
                token: TokenType::RBRACE,
                value: None,
                row: self.tok.row,
                col: self.tok.col,
            };
            self.bump();
            return Ok(());
        } else if self.current() == '=' {
            self.bump();
            if self.current() == '=' {
                self.tok = Token {
                    token: TokenType::COMPEQ,
                    value: None,
                    row: self.tok.row,
                    col: self.tok.col,
                };
                self.bump();
                return Ok(());
            }
            self.tok = Token {
                token: TokenType::EQ,
                value: None,
                row: self.tok.row,
                col: self.tok.col,
            };
            return Ok(());
        } else if self.current() == '!' {
            self.bump();
            if self.current() == '=' {
                self.tok = Token {
                    token: TokenType::COMPNE,
                    value: None,
                    row: self.tok.row,
                    col: self.tok.col,
                };
                self.bump();
                return Ok(());
            }
            self.tok = Token {
                token: TokenType::LOGNOT,
                value: None,
                row: self.tok.row,
                col: self.tok.col,
            };
            return Ok(());
        } else if self.current() == '>' {
            self.bump();
            if self.current() == '=' {
                self.tok = Token {
                    token: TokenType::COMPGE,
                    value: None,
                    row: self.tok.row,
                    col: self.tok.col,
                };
                self.bump();
                return Ok(());
            }
            self.tok = Token {
                token: TokenType::COMPGT,
                value: None,
                row: self.tok.row,
                col: self.tok.col,
            };
            return Ok(());
        } else if self.current() == '<' {
            self.bump();
            if self.current() == '=' {
                self.tok = Token {
                    token: TokenType::COMPLE,
                    value: None,
                    row: self.tok.row,
                    col: self.tok.col,
                };
                self.bump();
                return Ok(());
            }
            self.tok = Token {
                token: TokenType::COMPLT,
                value: None,
                row: self.tok.row,
                col: self.tok.col,
            };
            return Ok(());
        } else if self.current() == '&' {
            self.bump();
            if self.current() == '&' {
                self.tok = Token {
                    token: TokenType::COMPAND,
                    value: None,
                    row: self.tok.row,
                    col: self.tok.col,
                };
                self.bump();
                return Ok(());
            }
            self.tok = Token {
                token: TokenType::LOGAND,
                value: None,
                row: self.tok.row,
                col: self.tok.col,
            };
            return Ok(());
        } else if self.current() == '|' {
            self.bump();
            if self.current() == '|' {
                self.tok = Token {
                    token: TokenType::COMPOR,
                    value: None,
                    row: self.tok.row,
                    col: self.tok.col,
                };
                self.bump();
                return Ok(());
            }
            self.tok = Token {
                token: TokenType::LOGOR,
                value: None,
                row: self.tok.row,
                col: self.tok.col,
            };
            return Ok(());
        } else if self.current() == '^' {
            self.tok = Token {
                token: TokenType::LOGXOR,
                value: None,
                row: self.tok.row,
                col: self.tok.col,
            };
            self.bump();
            return Ok(());
        } else if self.current() == ':' {
            self.tok = Token {
                token: TokenType::COLON,
                value: None,
                row: self.tok.row,
                col: self.tok.col,
            };
            self.bump();
            return Ok(());
        } else if self.current() == '~' {
            self.tok = Token {
                token: TokenType::RANGE,
                value: None,
                row: self.tok.row,
                col: self.tok.col,
            };
            self.bump();
            return Ok(());
        } else if self.current() == '[' {
            self.tok = Token {
                token: TokenType::LBRACKET,
                value: None,
                row: self.tok.row,
                col: self.tok.col,
            };
            self.bump();
            return Ok(());
        } else if self.current() == ']' {
            self.tok = Token {
                token: TokenType::RBRACKET,
                value: None,
                row: self.tok.row,
                col: self.tok.col,
            };
            self.bump();
            return Ok(());
        } else if self.current() == ',' {
            self.tok = Token {
                token: TokenType::COMMA,
                value: None,
                row: self.tok.row,
                col: self.tok.col,
            };
            self.bump();
            return Ok(());
        } else if self.current() == '#' {
            while self.current() != '\n' && self.current() != '\0' {
                self.bump();
            }
            self.next_token()?;
            return Ok(());
        } else {
            return Err(LexerError::UnexpectedChar {
                expected: None,
                found: self.current(),
                row: self.tok.row,
                col: self.tok.col,
            });
        }
    }

    pub fn curr_tok(&self) -> Token {
        self.tok.clone()
    }

    pub fn curr_ch(&mut self) -> char {
        self.current()
    }
}
